# SPDX-License-Identifier: GPL-2.0-only
# Rewrites the six obstack.h macros that use lvalue-cast increments
# (*((T *) p)++ = v), which Clang rejects, into the equivalent
# store-then-advance form. Pure syntax change; no code generation impact.
# Apply from the gcc/ directory: patch -p1 < obstack.h.diff
--- a/obstack.h
+++ b/obstack.h
@@ -418,7 +418,8 @@
 ({ struct obstack *__o = (OBSTACK);					\
    if (__o->next_free + sizeof (void *) > __o->chunk_limit)		\
      _obstack_newchunk (__o, sizeof (void *));				\
-   *((void **)__o->next_free)++ = ((void *)datum);			\
+   *(void **)__o->next_free = ((void *)datum);				\
+   __o->next_free += sizeof (void *);					\
    (void) 0; })
 
 # define obstack_int_grow(OBSTACK,datum)				\
@@ -426,11 +427,12 @@
 ({ struct obstack *__o = (OBSTACK);					\
    if (__o->next_free + sizeof (int) > __o->chunk_limit)		\
      _obstack_newchunk (__o, sizeof (int));				\
-   *((int *)__o->next_free)++ = ((int)datum);				\
+   *(int *)__o->next_free = ((int)datum);					\
+   __o->next_free += sizeof (int);						\
    (void) 0; })
 
-# define obstack_ptr_grow_fast(h,aptr) (*((void **) (h)->next_free)++ = (void *)aptr)
-# define obstack_int_grow_fast(h,aint) (*((int *) (h)->next_free)++ = (int) aint)
+# define obstack_ptr_grow_fast(h,aptr) (*((void **) ((h)->next_free += sizeof (void *)) - 1) = (void *)aptr)
+# define obstack_int_grow_fast(h,aint) (*((int *) ((h)->next_free += sizeof (int)) - 1) = (int) aint)
 
 # define obstack_blank(OBSTACK,length)					\
 __extension__								\
@@ -537,8 +539,8 @@
    ? (_obstack_newchunk ((h), sizeof (int)), 0) : 0),			\
   (*((int *) (((h)->next_free+=sizeof(int))-sizeof(int))) = ((int) datum)))
 
-# define obstack_ptr_grow_fast(h,aptr) (*((char **) (h)->next_free)++ = (char *) aptr)
-# define obstack_int_grow_fast(h,aint) (*((int *) (h)->next_free)++ = (int) aint)
+# define obstack_ptr_grow_fast(h,aptr) (*((char **) ((h)->next_free += sizeof (char *)) - 1) = (char *) aptr)
+# define obstack_int_grow_fast(h,aint) (*((int *) ((h)->next_free += sizeof (int)) - 1) = (int) aint)
 
 # define obstack_blank(h,length)					\
 ( (h)->temp = (length),							\
